File: src/main/java/com/github/simiacryptus/aicoder/config/AppSettingsState.java
package com.github.simiacryptus.aicoder.config;

import com.github.simiacryptus.aicoder.openai.CompletionRequest;
import com.github.simiacryptus.aicoder.openai.EditRequest;
import com.github.simiacryptus.aicoder.openai.translate.TranslationRequest;
import com.github.simiacryptus.aicoder.openai.translate.TranslationRequestTemplate;
import com.intellij.openapi.application.ApplicationManager;
import com.intellij.openapi.components.PersistentStateComponent;
import com.intellij.openapi.components.State;
import com.intellij.openapi.components.Storage;
import com.intellij.util.xmlb.XmlSerializerUtil;
import com.jetbrains.rd.util.LogLevel;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.*;
import java.util.stream.Collectors;

/**
 * Supports storing the application settings in a persistent way.
 * The {@link State} and {@link Storage} annotations define the name of the data and the file name where
 * these persistent application settings are stored.
 */
@State(
        name = "org.intellij.sdk.settings.AppSettingsState",
        storages = @Storage("SdkSettingsPlugin.xml")
)
public class AppSettingsState implements PersistentStateComponent<AppSettingsState> {

    public @NotNull String apiBase = "https://api.openai.com/v1";
    public @NotNull String apiKey = "";
    public @NotNull String model_completion = "text-davinci-003";
    public @NotNull String model_edit = "text-davinci-edit-001";
    public int maxTokens = 1000;
    public double temperature = 0.1;
    public @NotNull String style = "";
    @SuppressWarnings("unused")
    public int tokenCounter = 0;
    private final @NotNull Map<String, Integer> mostUsedHistory = new HashMap<>();
    private final @NotNull List<String> mostRecentHistory = new ArrayList<>();
    public int historyLimit = 10;
    public @NotNull String humanLanguage = "English";
    public int maxPrompt = 5000;
    public @NotNull TranslationRequestTemplate translationRequestTemplate = TranslationRequestTemplate.XML;
    public @NotNull LogLevel apiLogLevel = LogLevel.Debug;
    public boolean devActions = false;

    public AppSettingsState() {
    }

    public static AppSettingsState getInstance() {
        return ApplicationManager.getApplication().getService(AppSettingsState.class);
    }

    public TranslationRequest createTranslationRequest() {
        return translationRequestTemplate.get(this);
    }

    public @NotNull CompletionRequest createCompletionRequest() {
        return new CompletionRequest(this);
    }

    public @NotNull EditRequest createEditRequest() {
        return new EditRequest(this);
    }

    @Nullable
    @Override
    public AppSettingsState getState() {
        return this;
    }

    @Override
    public void loadState(@NotNull AppSettingsState state) {
        XmlSerializerUtil.copyBean(state, this);
    }

    @Override
    public boolean equals(@Nullable Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        @NotNull AppSettingsState that = (AppSettingsState) o;
        if (maxTokens != that.maxTokens) return false;
        if (maxPrompt != that.maxPrompt) return false;
        if (Double.compare(that.temperature, temperature) != 0) return false;
        if (!Objects.equals(humanLanguage, that.humanLanguage)) return false;
        if (!Objects.equals(apiBase, that.apiBase)) return false;
        if (!Objects.equals(apiKey, that.apiKey)) return false;
        if (!Objects.equals(model_completion, that.model_completion)) return false;
        if (!Objects.equals(model_edit, that.model_edit)) return false;
        if (!Objects.equals(translationRequestTemplate, that.translationRequestTemplate)) return false;
        if (!Objects.equals(apiLogLevel, that.apiLogLevel)) return false;
        if (!Objects.equals(devActions, that.devActions)) return false;
        return Objects.equals(style, that.style);
    }

    @Override
    public int hashCode() {
        return Objects.hash(apiBase, apiKey, model_completion, model_edit, maxTokens, temperature, translationRequestTemplate, apiLogLevel, devActions, style);
    }

    public void addInstructionToHistory(@NotNull CharSequence instruction) {
        synchronized (mostRecentHistory) {
            mostRecentHistory.add(instruction.toString());
            while(mostRecentHistory.size() > historyLimit) {
                mostRecentHistory.remove(0);
            }
        }
        synchronized (mostUsedHistory) {
            mostUsedHistory.put(instruction.toString(), mostUsedHistory.getOrDefault(instruction, 0) + 1);
        }

        // If the instruction history is bigger than the history limit,
        // We'll make a set of strings to retain,
        // We'll sort the instruction history by its value,
        // And limit it to the history limit,
        // Then we'll map the entry key and collect it in a set,
        // Then we'll make a new hash set of the instruction history keys,
        // And remove all the ones we want to retain,
        // Then we'll remove all the ones we don't want to keep,
        // And that's how we'll make sure the instruction history is neat!
        if (mostUsedHistory.size() > historyLimit) {
            @NotNull List<CharSequence> retain = mostUsedHistory.entrySet().stream()
                    .sorted(Map.Entry.<String, Integer>comparingByValue().reversed())
                    .limit(historyLimit).map(Map.Entry::getKey).collect(Collectors.toList());
            @NotNull HashSet<CharSequence> toRemove = new HashSet<>(mostUsedHistory.keySet());
            toRemove.removeAll(retain);
            toRemove.removeAll(mostRecentHistory);
            toRemove.forEach(mostUsedHistory::remove);
        }
    }

    public @NotNull Set<String> getEditHistory() {
        return mostUsedHistory.keySet();
    }
}

File: src/main/kotlin/com/github/simiacryptus/aicoder/config/AppSettingsState.kt
package com.github.simiacryptus.aicoder.config

/**
 * Stores and manages plugin configuration settings.
 *
 * This class is responsible for persisting and retrieving the plugin's
 * configuration settings. It uses the IntelliJ Platform's persistence
 * framework to save settings across IDE restarts.
 */
import com.fasterxml.jackson.annotation.JsonIgnore
import com.github.simiacryptus.aicoder.util.PluginStartupActivity.Companion.addUserSuppliedModels
import com.intellij.openapi.application.ApplicationManager
import com.intellij.openapi.components.PersistentStateComponent
import com.intellij.openapi.components.State
import com.intellij.openapi.components.Storage
import com.intellij.openapi.util.io.FileUtil
import com.intellij.util.xmlb.XmlSerializerUtil
import com.simiacryptus.jopenai.models.APIProvider
import com.simiacryptus.jopenai.models.ImageModels
import com.simiacryptus.jopenai.models.OpenAIModels
import com.simiacryptus.skyenet.core.platform.AwsPlatform
import com.simiacryptus.util.JsonUtil
import org.slf4j.LoggerFactory
import java.io.File

@State(name = "org.intellij.sdk.settings.AppSettingsState", storages = [Storage("SdkSettingsPlugin.xml")])
data class AppSettingsState(
    var temperature: Double = 0.1,
    var smartModel: String = OpenAIModels.GPT4o.modelName,
    var fastModel: String = OpenAIModels.GPT4oMini.modelName,
    var mainImageModel: String = ImageModels.DallE3.modelName,
    var listeningPort: Int = 8081,
    var listeningEndpoint: String = "localhost",
    var humanLanguage: String = "English",
    var apiThreads: Int = 4,
    var apiBase: Map<String, String>? = mapOf("OpenAI" to "https://api.openai.com/v1"),
    var apiKey: Map<String, String>? = mapOf("OpenAI" to ""),
    var modalTasks: Boolean = false,
    var suppressErrors: Boolean = false,
    var apiLog: Boolean = false,
    var devActions: Boolean = false,
    var editRequests: Boolean = false,
    var disableAutoOpenUrls: Boolean = false,
    var storeMetadata: String? = null,
    var pluginHome: File = run {
        var logPath = System.getProperty("idea.plugins.path")
        if (logPath == null) {
            logPath = System.getProperty("java.io.tmpdir")
        }
        if (logPath == null) {
            logPath = System.getProperty("user.home")
        }
        File(logPath, "AICodingAsst")
    },
    var showWelcomeScreen: Boolean = true,
    var greetedVersion: String = "",
    var shellCommand: String = getDefaultShell(),
    var enableLegacyActions: Boolean = false,
    var executables: MutableSet<String> = mutableSetOf(),
    var recentArguments: MutableList<String> = mutableListOf(),
    val recentCommands: MutableMap<String, MRUItems> = mutableMapOf<String, MRUItems>(),
    var userSuppliedModels: MutableList<UserSuppliedModel> = mutableListOf(),
    var githubToken: String? = null,
    var googleApiKey: String? = null,
    var googleSearchEngineId: String? = null,
    var awsProfile: String? = null,
    var awsRegion: String? = null,
    var awsBucket: String? = null
) : PersistentStateComponent<SimpleEnvelope> {
    private var onSettingsLoadedListeners = mutableListOf<() -> Unit>()

    @JsonIgnore
    override fun getState(): SimpleEnvelope {
        val value = JsonUtil.toJson(this)
        return SimpleEnvelope(value)
    }

    fun getRecentCommands(id: String) = recentCommands.computeIfAbsent(id) { MRUItems() }

    override fun loadState(state: SimpleEnvelope) {
        state.value ?: return
        val fromJson = try {
            JsonUtil.fromJson(state.value!!, AppSettingsState::class.java)
        } catch (e: Exception) {
            log.warn("Error loading settings: ${state.value}", e)
            AppSettingsState()
        }
        XmlSerializerUtil.copyBean(fromJson, this)
        addUserSuppliedModels(fromJson.userSuppliedModels)
        recentCommands.clear()
        recentCommands.putAll(fromJson.recentCommands)
        notifySettingsLoaded()
    }

    fun addOnSettingsLoadedListener(listener: () -> Unit) {
        onSettingsLoadedListeners.add(listener)
    }

    private fun notifySettingsLoaded() {
        onSettingsLoadedListeners.forEach { it() }
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false
        other as AppSettingsState
        if (temperature != other.temperature) return false
        if (smartModel != other.smartModel) return false
        if (fastModel != other.fastModel) return false
        if (mainImageModel != other.mainImageModel) return false
        if (listeningPort != other.listeningPort) return false
        if (listeningEndpoint != other.listeningEndpoint) return false
        if (humanLanguage != other.humanLanguage) return false
        if (apiThreads != other.apiThreads) return false
        if (apiBase != other.apiBase) return false
        if (apiKey != other.apiKey) return false
        if (modalTasks != other.modalTasks) return false
        if (suppressErrors != other.suppressErrors) return false
        if (apiLog != other.apiLog) return false
        if (devActions != other.devActions) return false
        if (editRequests != other.editRequests) return false
        if (storeMetadata != other.storeMetadata) return false
        if (FileUtil.filesEqual(pluginHome, other.pluginHome)) return false
        if (recentCommands != other.recentCommands) return false
        if (showWelcomeScreen != other.showWelcomeScreen) return false
        if (greetedVersion != other.greetedVersion) return false
        if (mainImageModel != other.mainImageModel) return false
        if (enableLegacyActions != other.enableLegacyActions) return false
        if (executables != other.executables) return false
        //userSuppliedModels
        if (userSuppliedModels.toTypedArray().contentDeepEquals(other.userSuppliedModels.toTypedArray()).not()) return false
        if (googleApiKey != other.googleApiKey) return false
        if (googleSearchEngineId != other.googleSearchEngineId) return false
        if (githubToken != other.githubToken) return false
        if (awsProfile != other.awsProfile) return false
        if (awsRegion != other.awsRegion) return false
        if (awsBucket != other.awsBucket) return false
        return true
    }

    override fun hashCode(): Int {
        var result = temperature.hashCode()
        result = 31 * result + smartModel.hashCode()
        result = 31 * result + fastModel.hashCode()
        result = 31 * result + enableLegacyActions.hashCode()
        result = 31 * result + mainImageModel.hashCode()
        result = 31 * result + listeningPort
        result = 31 * result + listeningEndpoint.hashCode()
        result = 31 * result + humanLanguage.hashCode()
        result = 31 * result + apiThreads
        result = 31 * result + (apiBase?.hashCode() ?: 0)
        result = 31 * result + (apiKey?.hashCode() ?: 0)
        result = 31 * result + modalTasks.hashCode()
        result = 31 * result + suppressErrors.hashCode()
        result = 31 * result + apiLog.hashCode()
        result = 31 * result + devActions.hashCode()
        result = 31 * result + editRequests.hashCode()
        result = 31 * result + (storeMetadata?.hashCode() ?: 0)
        result = 31 * result + FileUtil.fileHashCode(pluginHome)
        result = 31 * result + recentCommands.hashCode()
        result = 31 * result + showWelcomeScreen.hashCode()
        result = 31 * result + greetedVersion.hashCode()
        result = 31 * result + mainImageModel.hashCode()
        result = 31 * result + enableLegacyActions.hashCode()
        result = 31 * result + executables.hashCode()
        result = 31 * result + userSuppliedModels.hashCode()
        result = 31 * result + (googleApiKey?.hashCode() ?: 0)
        result = 31 * result + (googleSearchEngineId?.hashCode() ?: 0)
        result = 31 * result + (githubToken?.hashCode() ?: 0)
        result = 31 * result + (awsProfile?.hashCode() ?: 0)
        result = 31 * result + (awsRegion?.hashCode() ?: 0)
        result = 31 * result + (awsBucket?.hashCode() ?: 0)
        return result
    }

    companion object {
        val log = LoggerFactory.getLogger(AppSettingsState::class.java)
        var auxiliaryLog: File? = null
        const val WELCOME_VERSION: String = "1.5.0"

        @JvmStatic
        val instance: AppSettingsState by lazy {
            ApplicationManager.getApplication()?.getService(AppSettingsState::class.java) ?: AppSettingsState()
        }

        fun String.imageModel(): ImageModels {
            return ImageModels.values().firstOrNull {
                it.modelName == this || it.name == this
            } ?: ImageModels.DallE3
        }

        fun getDefaultShell() = if (System.getProperty("os.name").lowercase().contains("win")) "powershell" else "bash"
    }

    data class UserSuppliedModel(
        var displayName: String = "",
        var modelId: String = "",
        var provider: APIProvider = APIProvider.OpenAI
    )
    var analyticsEnabled: Boolean = false
}